### ISSUE TITLE
provide message channel alternative for stream.Recv()

### ISSUE THREAD

**Issue by xh3b4sd** [grpc/grpc-go#847]
Hello folks. This is a feature request based on some inconvenience I experienced while playing around with `grpc-go`. I have some code that aims to stream messages bidirectionally. The server implementation redirects messages from and to go channels. Imagine the following code.

``` go
func (s *streamer) StreamText(stream pb.TextInterface_StreamTextServer) error {
  fail := make(chan error, 1)

  go func() {
    for {
      m, err := stream.Recv()
      if err != nil {
        fail <- maskAny(err)
        return
      }

      s.inputChan <- m
    }
  }()

  for {
    select {
    case <-stream.Context().Done():
      return stream.Context().Err()
        case err := <-fail:
      return err
    case m := <-s.ouputChan:
      err := stream.Send(m)
      if err != nil {
        return err
      }
    }
  }
}
```

The odd thing here is that I need to treat `stream.Recv` differently which also requires a new `fail` channel to redirect errors. What would be cool though is the following. Here I can handle all communication more elegant, because `stream.RecvChan` provides a go channel of incomming messages. The same would be cool to have for `stream.Send`. Thus it would also be cool to have `stream.SendChan`.

``` go
func (s *streamer) StreamText(stream pb.TextInterface_StreamTextServer) error {
  for {
    select {
    case <-stream.Context().Done():
      return stream.Context().Err()
    case m, err := <-stream.RecvChan():
      if err != nil {
        return err
      }
      s.inputChan <- m
    case m := <-s.ouputChan:
      err := stream.Send(m)
      if err != nil {
        return err
      }
    }
  }
}
```

I am aware that I should separate traffic from different directions into their own for loops. This here is to draw the picture. What am I missing? What do you guys think?

**Comment by mingrammer** [comment]
Any update?

**Comment by puellanivis** [comment]
A central problem here is that you cannot do a two destination receive and have the second value be an error. It will always be a `bool` that indicates whether the channel has closed or not.

So this code:
```
    case m, err := <-stream.RecvChan():
      if err != nil {
        return err
      }
      s.inputChan <- m
```

Will never work until/unless Golang expands how channel receives work.

It could be possible for `RecvChan()` to return a `chan struct { Message *pb.MyMessage, Error error }` and then you could have:
```
  case wrap := <-stream.RecvChan():
    if wrap.Error != nil {
      return err
    }
    s.inputChan <- wrap.Message
```

But this still isn’t “idiomatic”. The way streams work in gRPC makes a direct mapping onto native Go channels difficult no matter which way it is implemented.

**Comment by ItalyPaleAle** [comment]
> It could be possible for `RecvChan()` to return a `chan struct { Message *pb.MyMessage, Error error }`

There's actually lots of prior art for doing this, for example with go-etcd: the channel returns a message object, which might contain a value or an error

**Comment by dfawley** [comment]
It looks like the team never commented on this issue previously; we just applied some labels.  Sorry.  Our view:

This is not something we will be pursuing given our current API:

1. We don't want to add a second way of doing the same thing.

2. This would require a dedicated goroutine per-RPC for pushing to the recv channel and another for pulling from a send channel.  It might be worse than that: the generated code may also need its own pair of goroutines to convert from `interface{}` to the right proto message type (and if so, that would have a bad interaction with flow control, so we might also need to redesign the way registration works).

This could be something to consider if we design a v2 of the Go API, however, so we left it open with the "v2" label.

**Comment by ItalyPaleAle** [comment]
Thanks for responding @dfawley.

While I can respect your point 1, I don't think your point 2 should be a concern. In fact, while it's true that the library this way can avoid a dedicated goroutine for each gRPC, we end up having to implement it in our code, so the net effect is zero.

As you can see from the code @xh3b4sd shared above, as well as my ask on #3909 (which contains a similar ask, though I had a different issue), an additional goroutine per RPC (the `go` call within the goroutine that is already processing each RPC request) ends up being necessary in the users' code anyways. Sadly, doing it this way, the code just looks worse.

Given how big of a breaking change this would be, I have little hope that this will _replace_ the current API, as it would force lots of people to upgrade their code. I actually wouldn't advocate for it myself, as I'm the first one to say that breaking other people's code for the sake of doing it is not the nicest thing to do... But, I am hoping this could be considered (maybe for v1 too) as an _additional_ API  that is more idiomatic with the way Go is written...

Just my .02

**Comment by dfawley** [comment]
> we end up having to implement it in our code

* for some use cases.  There are others that don't require a dedicated goroutine.  E.g. the server may do interleaved send/recv operations.

> I have little hope that this will replace the current API, as it would force lots of people to upgrade their code

This violates semver in ways that we wouldn't allow, so it's a non-starter to delete the current API except in v2.

> more idiomatic with the way Go is written

It's important to realize this is an opinion and not an objective statement.  There are plenty of APIs in the standard packages that don't use channels when they would be possible, e.g. `io` or `http` don't use them at all.  In fact, channels are probably used more by the standard packages to signal events (e.g. context cancelation/timer firing) than to transmit data.  Consider some of these counter-arguments for reasons against using channels here: https://about.sourcegraph.com/go/idiomatic-go/#asynchronous-apis

**Comment by puellanivis** [comment]
I mean, I understand that we could just have the channel receive a struct that is both a message or an error. I called it out. But as I mentioned, this ends up not really really any cleaner, than just calling a function that returns either a message or an error. We end up hacking up a whole structure type just to permit us to avoid calling a function? 🤔 That doesn’t make sense to me.

I spent a really long time when I first started checking out gRPC and got excited by the streaming functions, that I would be able to do something with go channels that felt right… like `for result := range results { … }`, but in the end, due to the semantics of how the go stream works, none of them really had any benefits over just calling a function `Recv` and/or a function `Send`. The semantics here just don’t “fit”. I mean, we can force a square peg into a round hole by breaking either one, but then… have we really “solved” anything, or did we just spend a whole bunch of time beating out head against something, rather than just doing the simple thing and putting the square peg in the square hole?
